### PR TITLE
API Challenge's implementation

### PR DESCRIPTION
**Pull request by paul-albert**
# Description

- Finished implementation of the requested API endpoint.
- It was enough simple and generic, though had to keep in mind a lot of details.
- Tried to follow all instructions in `README.md`.

## Used technologies/tools

- Python 3.7 (.pyenv/venv)
- PyCharm
- Postman

## Elapsed time

~6h

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [x] Other

# How can this be tested?

- [x] Just do: `make unit-test` and `make coverage`, nothing complex
- [x] The tests are very generic, as they just check how does the API-route handle/answer to valid or invalid JSON-payloads
- [x] Purpose of the `pytest.ini`:
    - [x] setup proper paths to sources
    - [x] suppress the annoying warnings about deprecated JSON validator classes, though it's not mandatory
- [ ] Sure, the tests would be more complicated, e.g. additionally check if the customer data were stored properly, or some another edge cases

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and it is complete
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [ ] I have updated the changelog to include a description of my changes.

## Comments

- The "rodeo" :smile: was a good and a little bit simple challenge for me in general.
- Tried to follow good style of programming (e.g. singleton design pattern).
- Usage of the linter `black` was interesting for me, it's really cool thing, enjoyed of that, thanks.
- Of course, unit-tests could be more complicated and more full, especially for testing of saving for customer data.
- Also, wanted to try some parallelization for the API-endpoint, i.e. situation when API-routing receives a lot of requests. Possible solutions for that could be:
    - multithreading (though not sure due to the GIL)
    - multiprocessing (a bit better, but need to be careful with the possible data sharing between processes)
    - asynchronous requests (`asyncio` + `aiohttp` + `aiofiles`) - looks promising, but that's out of the challenge's borders
    - `celery` + queues (like as RabbitMQ/ActiveMQ/ZeroMQ/Redis or maybe even Apache Kafka) - that's the most optimal variant

### CODE REVIEW

File: .env.cli [hunk 1]
```diff
@@ -0,0 +1,2 @@
+VEHICLES_STORAGE_FOLDER=./vehicles_data
+MAX_VEHICLES_COUNT=10
```
Comment (markus-fuereder) [inline .env.cli:2]: 😎

File: json/vehicle-features.v1.example.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-  "vehicle": [
+  "vehicles": [
```
Comment (markus-fuereder) [inline json/vehicle-features.v1.example.json:2]: ✅

File: requirements.txt [hunk 3]
```diff
@@ -5,6 +5,9 @@ black==22.3.0
 click
 flask<=2.1.3
 flask_cors
+jsonschema==4.17.3
+python-dotenv<=0.21.1
+Werkzeug<=2.2.3
```
Comment (markus-fuereder) [inline requirements.txt:10]: Good choices 👍

File: src/core/challenge_api.py [hunk 4]
```diff
@@ -1,23 +1,46 @@
...
 
 api = Blueprint("challenge_api", __name__)
 
 
 @api.route("/challenge", methods=["POST"])
 @ApiDecorators.require_customer_id
-def vehicle_features_post(user_id: str):
+@ApiDecorators.validate_json_data("json/vehicle-features.v1.schema.json")
+@ApiDecorators.validate_uniqueness("vehicles", "id")
+@ApiDecorators.validate_max_count("vehicles", "id", int(getenv("MAX_VEHICLES_COUNT", default=10)))
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:17]: 🤩

File: src/core/challenge_api.py [hunk 5]
```diff
@@ -1,23 +1,46 @@
...
-    :param user_id: The id of the customer sending the request
-    :return: ???
+    :return:            Response from API-route
+    :rtype:             Response
     """
 
-    # TODO Implement the challenge
+    json_data = request.get_json()
+
+    storage = Storage(storage_folder=getenv("VEHICLES_STORAGE_FOLDER", default="./"))
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:34]: 🤩

File: src/utils/api_decorators.py [hunk 6]
```diff
@@ -1,15 +1,71 @@
...
+from utils.debug import debug
 
-class ApiDecorators:
-
-    MOCK_CUSTOMER_ID = str(uuid4())
 
+class ApiDecorators:
     @staticmethod
-    def require_customer_id(f):
+    def require_customer_id(f: Callable) -> Callable:
```
Comment (markus-fuereder) [inline src/utils/api_decorators.py:13]: Type annotations 🙌

File: src/utils/singleton.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
+from typing import Any, Dict, List, Tuple, Union
+
+
+class Singleton(type):
```
Comment (markus-fuereder) [inline src/utils/singleton.py:4]: 😎

File: src/utils/storage.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+        if json_data is None:
+            json_data = {}
+
+        user_id = secure_filename(user_id)
+        user_path = f"{self.storage_folder}/{user_id}".rstrip("/")
+        if os.path.exists(user_path) is False:
+            os.makedirs(user_path)
+
+        entity_id = secure_filename(entity_id)
```
Comment (markus-fuereder) [inline src/utils/storage.py:32]: ❤️ werkzeug

File: src/utils/validator.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
+import json
+from jsonschema import validate
+from jsonschema.exceptions import ValidationError
+import os
+from typing import Dict
+
+from utils.singleton import Singleton
+
+
+class Validator(object, metaclass=Singleton):
```
Comment (markus-fuereder) [inline src/utils/validator.py:10]: Very nice

**Review by markus-fuereder — COMMENTED** [review]
Solid submission. Ticked every of my boxes. Thanks for your effort.

I talked to Austyn she will schedule a google meets call.
Looking forward to get to know you and chat about the position. I think you would make a great fit.
